### PR TITLE
Add support for the symbolic string literal

### PR DESCRIPTION
**Pull request by NatashaWso2**
The characters followed by the " ' " in the symbolic string literal should start with a character in the range of  "A .. Z"  or "a .. z"  or with"_ ". The subsequent characters can have a combination of digits and characters ranging from "A .. Z"  and "a .. z".

The following characters are not allowed:
- AsciiChar := 0x0 .. 0x7F
- Unicode characters of category "Private Use" ranging from: 0xE000 .. 0xF8FF | 0xF0000 .. 0xFFFFD | 0x100000 .. 0x10FFFD
- Unicode whitespace chars : 0x200E | 0x200F | 0x2028 | 0x2029
- Unicode characters with property Pattern_Syntax=True

The following are valid: 
```ballerina
// Variable definition
string str = 'HelloWorld;
string str = '_HelloWorld;
string str = 'HelloWorld123;

// Used in maps
map addrMap = { number: "No.20", road: 'Mount_Lavinia, country: 'SriLanka };
addrMap['postalCode] = 'PO00300;
string country = <string>addrMap['country];

// Used in tuples
 (int, string, string) a = (10, 'John_Snow, 'UK);

// Used in json
json p = { fname: 'John, lname: 'Stallone, 'age: 30 };
json lastName = p['lname];
p['age] = 66;

// Used in objects
Person p2 = new('Adam, name = 'Adam, lastname = 'Page);

// Used in records
Student stu = { name: 'AdamPage, age: 17};
string name = stu['name];
```
The following are not valid:
```ballerina
string var1 = 'Hello World;
string var2 = 'Hello$;
string var3 = 'U+001E;
string var4 = 'Hello\nWorld;
string var6 = 'Hello󰀇;
string var6 = 'Hello��;;

// Used in maps
map addrMap = { road: 'Mount Lavinia, country: 'Sri Lanka };
addrMap['country sl] = 'P0300;

// Used in records
Student stu = { name: 'Adam Page, age: 17};
string name = stu['name$];

// Used in arrays
string [] stringArr = ['hello world];

// Used in json
json p = { name: 'John\uFFFFStallone };
```

## Automation tests
 - Unit tests 
   > Added both positive and negative test cases

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangSymbolicStringLiteral.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.wso2.ballerinalang.compiler.tree.expressions;
+
+import org.ballerinalang.model.tree.NodeKind;
+import org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor;
+
+/**
+ * @since 0.982
```
Comment (vinok88) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangSymbolicStringLiteral.java]: 0.982 is already released right?

Comment (NatashaWso2) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangSymbolicStringLiteral.java]: Yes it is already released. Added the correct version with 75154ec. Thanks :)

File: tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-negative-test.bal [hunk 2]
```diff
@@ -0,0 +1,25 @@
+type Student record {
```
Comment (hemikak) [inline tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-negative-test.bal]: Need license header.

File: tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal [hunk 3]
```diff
@@ -0,0 +1,225 @@
+import ballerina/crypto;
```
Comment (hemikak) [inline tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal]: Need license header

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java [hunk 4]
```diff
@@ -576,7 +576,8 @@ void visitObjectAttachedFunctionParamSymbol(BLangVariable variable, SymbolEnv in
         if (variable.expr == null) {
             return;
         }
-        if (variable.expr.getKind() != NodeKind.LITERAL) {
+        if (variable.expr.getKind() != NodeKind.LITERAL && variable.expr.getKind() != NodeKind.
+                SYMBOLIC_STRING_LITERAL) {
```
[the remark this answers is not in the document]
Comment (NatashaWso2) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java]: Improved formatting with fe2a22e

File: tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal (hunk 3 above)
Comment (NatashaWso2) [reply tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal]: Added license header with fe2a22e

File: tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-negative-test.bal (hunk 2 above)
Comment (NatashaWso2) [reply tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-negative-test.bal]: Added license header with fe2a22e

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java [hunk 5]
```diff
@@ -2174,25 +2174,33 @@ public void exitSimpleLiteral(BallerinaParser.SimpleLiteralContext ctx) {
         Object value;
         BallerinaParser.IntegerLiteralContext integerLiteralContext = ctx.integerLiteral();
         if (integerLiteralContext != null && (value = getIntegerLiteral(ctx, ctx.integerLiteral())) != null) {
-            this.pkgBuilder.addLiteralValue(pos, ws, TypeTags.INT, value);
+            this.pkgBuilder.addLiteralValue(pos, ws, TypeTags.INT, value,
+                                            getOriginalIntegerValue(integerLiteralContext));
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: is ctx.getText() giving the orginal string value. ?
```suggestion
                                            ctx.getText());
```

Comment (NatashaWso2) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: Yes ctx.getText() gives the original string value. Fixed it with 305aa04. Thanks :)

Comment (NatashaWso2) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java]: Sorry seems like the change was not committed properly :( Updated the PR with cee2b10

File: tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+string var1 = 'abcd;
+
+public string var2;
+
+@final
+public string var3 = 'efgh;
+
+@readonly
+public string var4 = 'ijkl;
+
```
Comment (Kishanthan) [inline tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal]: Need test cases combining quoted string with symbolic string

Comment (NatashaWso2) [reply tests/ballerina-unit-test/src/test/resources/test-src/types/string/symbolic-string-test.bal]: Added test cases with f097e79

**Comment by NatashaWso2** [comment]
@Kishanthan sorry I had not pushed my changes before :/  Pushed the generated parser with ddc0141

**Pull request merged**